datalad_next/itertools/reroute.py

```diff
@@ -0,0 +1,232 @@
+)
+
+
+__all__ = ['dont_process', 'join_with_list', 'route_in', 'route_out']
+
+
+dont_process = object()
+
+
+def route_out(iterable: Iterable,
+              data_store: list,
+              splitter: Callable[[Any], tuple[Any, Any | None]],
+              ) -> Generator:
+    """ Route data around the consumer of this iterable
+
+    :func:`route_out` allows its user to:
+
+     1. store data that is received from an iterable,
+     2. determine whether this data should be yielded to a consumer of
+        ``route_out``, by calling :func:`splitter`.
+
+    To determine which data is to be yielded to the consumer and which data
+    should only be stored but not yielded, :func:`route_out` calls
+    :func:`splitter`. :func:`splitter` is called for each element of the input
+    iterable, with the element as sole argument. The function should return a
+    tuple of two elements. The first element is the data that is to be
+    yielded to the consumer. The second element is the data that is to be
+    stored in the list ``data_store``. If the first element of the tuple is
+    ``datalad_next.itertools.dont_process``, no data is yielded to the
+    consumer.
+
+    :func:`route_in` can be used to combine data that was previously
+    stored by :func:`route_out` with the data that is yielded by
+    :func:`route_out` and with the data the was not processed, i.e. not yielded
+    by :func:`route_out`.
+
+    The elements yielded by :func:`route_in` will be in the same
+    order in which they were passed into :func:`route_out`, including the
+    elements that were not yielded by :func:`route_out` because :func:`splitter`
+    returned ``dont_process`` in the first element of the result-tuple.
+
+    The combination of the two functions :func:`route_out` and :func:`route_in`
+    can be used to "carry" additional data along with data that is processed by
+    iterators. And it can be used to route data around iterators that cannot
+    process certain data.
+
+    For example, a user has an iterator to divide all number in a list by the
+    number ``2``. The user wants the iterator to process all numbers in a list,
+    except from zeros, In this case :func:`route_out` and :func:`route_in` can
+    be used as follows:
+
+    .. code-block:: python
+
+        from math import nan
+        from datalad_next.itertools import route_out, route_in, dont_process
+
+        def splitter(data):
+            # if data == 0, return `dont_process` in the first element of the
+            # result tuple to indicate that route_out should not yield this
+            # element to its consumer
+            return (dont_process, [data]) if data == 0 else (data, [data])
+
+        def joiner(processed_data, stored_data):
+            #
+            return nan if processed_data is dont_process else processed_data
+
+        numbers = [0, 1, 0, 2, 0, 3, 0, 4]
+        store = list()
+        r = route_in(
+            map(
+                lambda x: x / 2.0,
```

Comment:
The example feels a bit weird. 0/2 is 0 not nan. I think the order of variable should be switched here.

Reply:
Yes, of course. Thanks for catching that. Done in 9edab7517295a2a8007b562933c5ecb6910fac41